### PR TITLE
Hot Backup Automation

### LINKED ISSUE #3905: Question: HotBackup test pipeline
For Hot Backup feature we have test cases like this:

```
        @Setup:
        Populate the system with hosts/repos/contents
        @Steps:
        1. Run full backup (base)
        2. Make config change c1
        3. Run incremental backup ib1
        4. Restore base backup, verify c1 config doesnt not exist
        5. restore ib1, veirfy c1 config does exist
        6. Make config change c2
        7. run katello-backup <path_to_ib1> --incremental
        @Assert: Backup "ib1" is backed up.
```

Online backup can run along with the system (no need to shutdown or restart the system) but it is time consuming, also we can have race conditions if backup occurs along with other automation tests.

So I think we need a separated pipeline for sys/backup tests, reasons:
1. we do not want it to take longtime so we are not going to backup a heavily full system
2. we want to assert incremental backups so ensure parallel tests will not compromise the assertions
3. if something went wrong we want to have a clean debug and log

> NOTE: Backup tests are all `run_in_one_thread`  
## **Solutions**:

**Put at the end of existing pipeline.**
1. install pytest-order plugin
2. make backup tests to be the latest to run after some tier test ends
3. record separate artifacts on backup class teardown (system logs)

Pros: we have a system populated by previous tests. 
Cons: Backup errors will fail that tier, we can have more entities than needed taking a longtime to backup.

---

**Create a new pipeline only for sys/backup tests**
1. It can be triggered after tier4, we can have a `pytest.mark.sys`
2. Starts with clean system
3. Populate the system only with sufficient number of important entities
4. Artifacts will be separated anyway
5. It is a ONE THREAD pipeline

Pros: all the above
Cons: One more pipeline to maintain

---

Ideas?

### PR DESCRIPTION
**Pull request by rochacbruno**
- Automated 3 cases
- Waiting to resolve #3905 
- Waiting to clarify FAQs from feature

[6.3 feature is in "back burner" right now, but I want to merge those 3 cases automated and then in future when we got back to 6.3 I can continue on this]

### CODE REVIEW

**Comment by rochacbruno** [comment]
I want to merge this in to master to put in improvement I made in SSH context manager and also avoid conflicts in future.

```
$ py.test -v tests/foreman/sys/ -m "not stubbed"
====== test session starts =====
collected 9 items 

tests/foreman/sys/test_hot_backup.py::HotBackupTestCase::test_positive_backup_with_existing_directory PASSED
tests/foreman/sys/test_hot_backup.py::HotBackupTestCase::test_positive_incremental PASSED
tests/foreman/sys/test_hot_backup.py::HotBackupTestCase::test_positive_skip_pulp PASSED

===== 6 tests deselected by "-m 'not stubbed'" ======
===== 3 passed, 6 deselected in 47.29 seconds =====

```

@pgagne @abalakh @oshtaier @renzon can you review/ACK/merge?

File: tests/foreman/sys/test_hot_backup.py [hunk 1]
```diff
@@ -14,42 +14,46 @@
...
-        @Assert: The backup and restore succeed.
 
-        @CaseAutomation: notautomated
+        @Assert: The backup is successfully created in existing folder and
+        contains all the default files needed to restore.
 
-        @CaseLevel: System
         """
+        with get_connection() as connection:
+            connection.run('mkdir /tmp/backup1')
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Should the name of the directory be a random string? This way one can run the same test over and over or in parallel and avoid using previous artifacts?

File: tests/foreman/sys/test_hot_backup.py [hunk 2]
```diff
@@ -14,42 +14,46 @@
...
+            result = connection.run(
+                'katello-backup --online-backup /tmp/backup1', 'plain')
+            self.assertEqual(result.return_code, 0)
+            expected_msg = (
+                'BACKUP Complete, contents can be found in: /tmp/backup1'
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: If you use a random name, then your expected message would have to be dynamic too.

File: tests/foreman/sys/test_hot_backup.py [hunk 3]
```diff
@@ -14,42 +14,46 @@
...
+            )
+            self.assertIn(expected_msg, result.stdout)
+            files = connection.run('ls /tmp/backup1', 'list')
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Same comment about random names.

File: tests/foreman/sys/test_hot_backup.py [hunk 4]
```diff
@@ -59,79 +63,125 @@ def test_positive_directory_created(self):
...
+        1. Run backup with --skip-pulp option to /tmp/bck-no-pulp
+        2. List contents of the destination
+
+        @Assert: ``/tmp/bck-no-pulp`` is created and pulp related files is not
+        present.
 
-        @CaseLevel: System
         """
+        with get_connection() as connection:
+            connection.run('rm -rf /tmp/bck-no-pulp')
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Use random names?

File: tests/foreman/sys/test_hot_backup.py [hunk 5]
```diff
@@ -59,79 +63,125 @@ def test_positive_directory_created(self):
...
+        files in destination.
+
+        @id: 0f0c1915-dd07-4571-9b05-e0778efc85b2
+
+        @Steps:
+
+        1. Run backup with --skip-pulp option to /tmp/bck-no-pulp
+        2. List contents of the destination
+
+        @Assert: ``/tmp/bck-no-pulp`` is created and pulp related files is not
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: s/is not/are not

File: tests/foreman/sys/test_hot_backup.py [hunk 6]
```diff
@@ -59,79 +63,125 @@ def test_positive_directory_created(self):
...
         @Assert: Backup "ib1" is backed up.
 
-        @CaseAutomation: notautomated
-
-        @CaseLevel: System
         """
+        with get_connection() as connection:
+            # run full backup
+            result = connection.run(
+                'katello-backup --online-backup /tmp/b1', 'plain')
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Use random names?

File: tests/foreman/sys/test_hot_backup.py [hunk 7]
```diff
@@ -59,79 +63,125 @@ def test_positive_directory_created(self):
...
-
-        Populate a Satellite with hosts/repos/contents
-
         @Steps:
 
-        1. Restore online backup
-        2. Restore offline backup
-        3. Restore online incremenetal backup
-        4. Restore offline incremenetal backup
+        1. Create backup on previous version of satellite
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: s/on/from?

File: tests/foreman/sys/test_hot_backup.py [hunk 8]
```diff
@@ -59,79 +63,125 @@ def test_positive_directory_created(self):
...
+            # restore /tmp/b1 and assert change 1 is not there
+
+            # restore /tmp/ib1 and assert change 1 is there
 
     @tier3
     @stubbed()
-    def test_positive_restore_fresh_install(self):
-        """Restore to a clean Satellite with no config
+    def test_positive_restore_after_update(self):
+        """Restore a backup in updated version of satellite
```
Comment (omaciel) [inline tests/foreman/sys/test_hot_backup.py]: Restore from a backup?

**Comment by rochacbruno** [comment]
@omaciel changed done! 1 test skipped due to a reported bug using `--incremental` 3 tests passing.

**Comment by pondrejk** [comment]
ACK

**Pull request merged**